src/dmd/backend/dt.d

```diff
@@ -5,75 +5,616 @@
+            dtn = dtnext;
+        }
+        dtn.DTnext = dt_freelist;
+        dt_freelist = dt;
+    }
+}
+
+/*********************************
+ * Free free list.
+ */
+
+void dt_term()
+{
+static if (0 && TERMCODE)
+{
+    dt_t *dtn;
+
+    while (dt_freelist)
+    {   dtn = dt_freelist.DTnext;
+        mem_ffree(dt_freelist);
+        dt_freelist = dtn;
+    }
+}
+}
+
+dt_t **dtend(dt_t **pdtend)
+{
+    while (*pdtend)
+        pdtend = &((*pdtend).DTnext);
+    return pdtend;
+}
+
+
+/*********************************
+ */
+void dtpatchoffset(dt_t *dt, uint offset)
+{
+    dt.DToffset = offset;
+}
+
+/**************************
+ * Make a common block for s.
+ */
+
+void init_common(Symbol *s)
 {
-    void dt_free(dt_t*);
-    void dtpatchoffset(dt_t *dt, uint offset);
-    bool dtallzeros(const(dt_t)* dt);
-    bool dtpointers(const(dt_t)* dt);
-    void dt2common(dt_t **pdt);
-    uint dt_size(const(dt_t)* dtstart);
+    //printf("init_common('%s')\n", s.Sident);
+
+    uint size = cast(uint)type_size(s.Stype);
+    if (size)
+    {
+        dt_t *dt = dt_calloc(DT_common);
+        dt.DTazeros = size;
+        s.Sdt = dt;
+    }
 }
 
-extern (C++) class DtBuilder
+/**********************************
+ * Compute size of a dt
+ */
+
+uint dt_size(const(dt_t)* dtstart)
+{
+    uint datasize = 0;
+    for (auto dt = dtstart; dt; dt = dt.DTnext)
+    {
+        switch (dt.dt)
+        {
+            case DT_abytes:
+                datasize += size(dt.Dty);
+                break;
+            case DT_ibytes:
+                datasize += dt.DTn;
+                break;
+            case DT_nbytes:
+                datasize += dt.DTnbytes;
+                break;
+            case DT_azeros:
+                datasize += dt.DTazeros;
+                break;
+            case DT_common:
+                break;
+            case DT_xoff:
+            case DT_coff:
+                datasize += size(dt.Dty);
+                break;
+            default:
+                debug printf("dt = %p, dt = %d\n",dt,dt.dt);
+                assert(0);
+        }
+    }
+    return datasize;
+}
+
+/************************************
+ * Return true if dt is all zeros.
+ */
+
+bool dtallzeros(const(dt_t)* dt)
+{
+    return dt.dt == DT_azeros && !dt.DTnext;
+}
+
+/************************************
+ * Return true if dt contains pointers (requires relocations).
+ */
+
+bool dtpointers(const(dt_t)* dtstart)
+{
+    for (auto dt = dtstart; dt; dt = dt.DTnext)
+    {
+        switch (dt.dt)
+        {
+            case DT_abytes:
+            case DT_xoff:
+            case DT_coff:
+                return true;
+
+            default:
+                break;
+        }
+    }
+    return false;
+}
+
+/***********************************
+ * Turn DT_azeros into DTcommon
+ */
+
+void dt2common(dt_t **pdt)
+{
+    assert((*pdt).dt == DT_azeros);
+    (*pdt).dt = DT_common;
+}
+
+/**********************************************************/
+
+struct DtBuilder
```

Comment:
Why was this changed to a struct?

Reply:
presumably because it is no longer used from C++. Why not a final class, I'm not sure.

Reply:
It goes against Walter's rules about translating code, that's why I was asking.

Reply:
Good question. Dunno.

Reply:
Classes require TypeInfo, which is not allowed in -betterC. This needs to be better, but for the time being we are stuck with the bootstrap compiler as it is.

Reply:
Also, .di files do not work with -betterC, even with structs, so I had to abandon that as well.

Reply:
Fair enough.